README.md

# Card Banner
It is simple to add a banner to any widget using card banner.

## Features

- Highly customizable.
- Multi-platform support.


## Getting started

Add this line to your pubspec.yaml under the dependencies:
```dart
dependencies:
    card_banner: <latest-version>
```
alternatively, you can use this command:
```dart
flutter pub add card_banner
```
Import statement
```dart
import 'package:card_banner/card_banner.dart'
```
## Screenshot

<img src="https://user-images.githubusercontent.com/45100807/205489258-8d06faa3-c101-45ba-a33c-4bb24073de24.png" width="280" height="600">



## Usage

`Simple example:` Cover the card widget with the card banner and provide the required text. It's done

```dart
class MyBanner extends StatelessWidget {
  const MyBanner({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: const Color.fromARGB(255, 221, 222, 245),
      body: ListView.builder(
        padding: const EdgeInsets.symmetric(horizontal: 10),
        itemCount: 8,
        itemBuilder: (context, index) {
          return CardBanner(
            text: "50% Off",
            child: foodCard(),
          );
        },
      ),
    );
  }

  Card foodCard() {
    return Card(
      child: ListTile(
        contentPadding: EdgeInsets.zero,
        title: Image.network(
          "https://images.pexels.com/photos/461198/pexels-photo-461198.jpeg",
          fit: BoxFit.cover,
        ),
        subtitle: const Padding(
          padding: EdgeInsets.all(12.0),
          child: Text(
            "Nulla cillum qui elit adipisicing culpa ad laboris amet laboris velit. Pariatur sit tempor nostrud non duis commodo laboris. Consequat incididunt laborum qui dolore in velit consequat mollit amet. Proident eu et sint veniam sunt fugiat quis fugiat sunt pariatur do enim in.",
            overflow: TextOverflow.ellipsis,
            maxLines: 3,
          ),
        ),
      ),
    );
  }
}

```

## Additional information

Check the [example project](https://github.com/abhishek-900/card_banner/tree/master/example) on github for advanced usage.


CHANGELOG.md
## 0.0.1
* Added card banner widget and example project.


lib/src/card_banner_view.dart
import 'package:flutter/material.dart';

enum CardBannerPosition {
  TOPLEFT,
  TOPRIGHT,
  BOTTOMLEFT,
  BOTTOMRIGHT,
}

class CardBanner extends StatelessWidget {
  const CardBanner({
    super.key,
    required this.child,
    required this.text,
    this.position = CardBannerPosition.TOPLEFT,
    this.color = Colors.blue,
    this.edgeColor = const Color.fromARGB(255, 7, 86, 150),
    this.radius = 4,
    this.padding = 4.0,
    this.edgeSize = 6,
    this.textStyle = const TextStyle(fontSize: 14, color: Colors.white),
  });

  final Widget child;
  final String text;
  final Color color;
  final Color edgeColor;
  final double edgeSize;
  final double radius;
  final double padding;
  final CardBannerPosition position;
  final TextStyle textStyle;

  @override
  Widget build(BuildContext context) {
    return Stack(
      clipBehavior: Clip.none,
      children: [
        child,
        Positioned.fill(
          right: position.toString().contains("RIGHT") ? -6 : null,
          left: position.toString().contains("LEFT") ? -6 : null,
          top: position.toString().contains("TOP") ? 20 : null,
          bottom: position.toString().contains("BOTTOM") ? 20 : null,
          child: CustomPaint(
            painter: DesignCardBanner(edgeColor, position, edgeSize),
            child: Container(
              padding: EdgeInsets.all(padding),
              decoration: BoxDecoration(
                borderRadius: BorderRadius.only(
                  topLeft: Radius.circular(radius),
                  topRight: Radius.circular(radius),
                  bottomLeft: position.toString().contains("LEFT")
                      ? Radius.zero
                      : Radius.circular(radius),
                  bottomRight: position.toString().contains("RIGHT")
                      ? Radius.zero
                      : Radius.circular(radius),
                ),
                color: color,
              ),
              child: Text(text, style: textStyle),
            ),
          ),
        ),
      ],
    );
  }
}

class DesignCardBanner extends CustomPainter {
  DesignCardBanner(this.color, this.position, this.edgeSize);
  final Color color;
  final CardBannerPosition position;
  final double edgeSize;
  @override
  void paint(Canvas canvas, Size size) {
    if (position.toString().contains("LEFT")) {
      Paint paint = Paint()..color = color;
      var path = Path();
      path.moveTo(0, 0);
      path.lineTo(0, size.height);
      path.lineTo(10, size.height + edgeSize);
      path.lineTo(10, size.height);
      path.lineTo(0, size.height);
      path.close();
      canvas.drawPath(path, paint);
    } else {
      Paint paint = Paint()..color = color;
      var path = Path();
      path.moveTo(size.width, size.height);

      path.lineTo(size.width - 10, size.height + edgeSize);
      path.lineTo(size.width - 10, size.height);
      path.lineTo(size.width, size.height);
      path.close();
      canvas.drawPath(path, paint);
    }
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return false;
  }
}


example/lib/main.dart
import 'package:card_banner/card_banner.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Card Banner',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyBanner(),
    );
  }
}

class MyBanner extends StatelessWidget {
  const MyBanner({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: const Color.fromARGB(255, 221, 222, 245),
      body: ListView.builder(
        padding: const EdgeInsets.symmetric(horizontal: 10),
        itemCount: 8,
        itemBuilder: (context, index) {
          return CardBanner(
            text: "50% Off",
            child: foodCard(),
          );
        },
      ),
    );
  }

  Card foodCard() {
    return Card(
      child: ListTile(
        contentPadding: EdgeInsets.zero,
        title: Image.network(
          "https://images.pexels.com/photos/842571/pexels-photo-842571.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2",
          height: 200,
          fit: BoxFit.cover,
        ),
        subtitle: const Padding(
          padding: EdgeInsets.all(12.0),
          child: Text(
            "Nulla cillum qui elit adipisicing culpa ad laboris amet laboris velit. Pariatur sit tempor nostrud non duis commodo laboris. Consequat incididunt laborum qui dolore in velit consequat mollit amet. Proident eu et sint veniam sunt fugiat quis fugiat sunt pariatur do enim in.Qui proident veniam aute commodo consectetur id ipsum. Laborum occaecat enim cupidatat deserunt et reprehenderit tempor ipsum elit. Proident enim ullamco aliqua veniam cupidatat mollit voluptate dolore Lorem adipisicing velit. Culpa magna occaecat ipsum ea aliquip in quis sunt cillum ad deserunt nostrud ex non.",
            overflow: TextOverflow.ellipsis,
            maxLines: 3,
          ),
        ),
      ),
    );
  }
}
